airbyte-integrations/connectors/source-airtable/source_airtable/schema_helpers.py

```diff
@@ -3,35 +3,115 @@
+    "button": SchemaTypes.string,
+    "checkbox": SchemaTypes.boolean,
+    "singleCollaborator": SchemaTypes.string,
+    "count": SchemaTypes.number,
+    "createdBy": SchemaTypes.string,
+    "createdTime": SchemaTypes.datetime,
+    "currency": SchemaTypes.number,
+    "email": SchemaTypes.string,
+    "date": SchemaTypes.date,
+    "dateTime": SchemaTypes.datetime,
+    "duration": SchemaTypes.number,
+    "lastModifiedBy": SchemaTypes.string,
+    "lastModifiedTime": SchemaTypes.datetime,
+    "multipleRecordLinks": SchemaTypes.array_with_strings,
+    "multilineText": SchemaTypes.string,
+    "multipleCollaborators": SchemaTypes.array_with_strings,
+    "multipleSelects": SchemaTypes.array_with_strings,
+    "number": SchemaTypes.number,
+    "percent": SchemaTypes.number,
+    "phoneNumber": SchemaTypes.string,
+    "rating": SchemaTypes.number,
+    "richText": SchemaTypes.string,
+    "singleLineText": SchemaTypes.string,
+    "externalSyncSource": SchemaTypes.string,
+    "url": SchemaTypes.string,
+}
+
+# returns the `array of Any` where Any is based on Simple Types.
+# the final array is fulled with some simple type.
+COMPLEX_AIRTABLE_TYPES: Dict = {
+    "formula": SchemaTypes.array_with_any,
+    "lookup": SchemaTypes.array_with_any,
+    "multipleLookupValues": SchemaTypes.array_with_any,
+    "rollup": SchemaTypes.array_with_any,
+}
+
+
 class SchemaHelpers:
     @staticmethod
     def clean_name(name_str: str) -> str:
         return name_str.replace(" ", "_").lower().strip()
 
     @staticmethod
     def get_json_schema(table: Dict[str, Any]) -> Dict[str, str]:
-        fields = table.get("fields", {})
-        properties = {
-            "_airtable_id": {"type": ["null", "string"]},
-            "_airtable_created_time": {"type": ["null", "string"]},
+        properties: Dict = {
+            "_airtable_id": SchemaTypes.string,
+            "_airtable_created_time": SchemaTypes.string,
         }
 
+        fields: Dict = table.get("fields", {})
         for field in fields:
-            field_name = SchemaHelpers.clean_name(field.get("name"))
-            properties[field_name] = {"type": ["null", "string"]}
+            name: str = SchemaHelpers.clean_name(field.get("name"))
+            original_type: str = field.get("type")
+            options: Dict = field.get("options", {})
+            exec_type: str = options.get("result", {}).get("type") if options else None
+            field_type: str = exec_type if exec_type else original_type
+
+            # choose the JsonSchema Type for known Airtable Types
+            if original_type in COMPLEX_AIRTABLE_TYPES.keys():
+                complex_type = deepcopy(COMPLEX_AIRTABLE_TYPES.get(original_type))
+                # process arrays with values
+                if complex_type == SchemaTypes.array_with_any:
+                    complex_type["items"] = deepcopy(SIMPLE_AIRTABLE_TYPES.get(field_type))
+                properties.update(**{name: complex_type})
```

Comment:
I'm a little muddy on how the `result` is working - but I am wondering if we can do anything to enforce the type of other `options` returned by some complex types. I'm looking for example at [formula](https://airtable.com/developers/web/api/field-model#formula) and handling the boolean in the options. Maybe its all going to the same column in a nested format so it doesn't matter anyway?

Reply:
The `COMPLEX` type is something that I've `overthought` to give the explanation of  the next structure due  to Airtable Types:
- COMPLEX could be one of the:
or `[ string. string, ... ] `
or `[ [list], [list] ] `
or `[ [string, string, etc], [string, string], etc ]`

It's basically the Array with some other `SIMPLE` type inside. It circles back and forth to get the complex type first,  then looks for the simple type that the complex type contains inside. All the information is stored inside the response from the schema API request, in the `options` > `result` object that points to the exact type that column handles. The path is always the same for all  complex  types, if something is wrong it will not break but will be cast as a `string`

Reply:
Also, here we have a deal with `array_with_any` type, so we handle them specifically here:
```python
# process arrays with values
 if complex_type == SchemaTypes.array_with_any:
    complex_type["items"] = deepcopy(SIMPLE_AIRTABLE_TYPES.get(field_type))
```

Once there are some `array_with_any` and  we have the  `object` inside, instead  of `array`, it will be handled  like:
```python
# process arrays with values
   if complex_type == SchemaTypes.array_with_any_object:
      complex_type["properties"] = deepcopy(SIMPLE_AIRTABLE_TYPES.get(field_type)) // assuming we are dealing with ready to go field names 
 ```